src/Microsoft.AspNet.Mvc.Core/Internal/FilterCache.cs

```diff
@@ -0,0 +1,211 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using Microsoft.AspNet.Mvc.Abstractions;
+using Microsoft.AspNet.Mvc.Filters;
+using Microsoft.AspNet.Mvc.Infrastructure;
+
+namespace Microsoft.AspNet.Mvc.Internal
+{
+    public class FilterCache
+    {
+        private readonly IFilterMetadata[] EmptyFilterArray = new IFilterMetadata[0];
+
+        private readonly IActionDescriptorCollectionProvider _collectionProvider;
+        private readonly IFilterProvider[] _filterProviders;
+
+        private InnerCache _currentCache;
+
+        public FilterCache(
+            IActionDescriptorCollectionProvider collectionProvider,
+            IEnumerable<IFilterProvider> filterProviders)
+        {
+            _collectionProvider = collectionProvider;
+            _filterProviders = filterProviders.OrderBy(item => item.Order).ToArray();
+        }
+
+        private InnerCache CurrentCache
+        {
+            get
+            {
+                var current = Volatile.Read(ref _currentCache);
+                var actionDescriptors = _collectionProvider.ActionDescriptors;
+
+                if (current == null || current.Version != actionDescriptors.Version)
+                {
+                    current = new InnerCache(actionDescriptors.Version);
+                    Volatile.Write(ref _currentCache, current);
+                }
+
+                return current;
+            }
+        }
+
+        public IFilterMetadata[] GetFilters(ActionContext actionContext)
+        {
+            var cache = CurrentCache;
+            var actionDescriptor = actionContext.ActionDescriptor;
+
+            CacheEntry entry;
+            if (cache.Entries.TryGetValue(actionDescriptor, out entry))
```

Comment:
What happens this method is invoked an outdated (old version) `ActionDescriptor`?